### ISSUE TITLE
Some questions about the loss funtion?

### ISSUE THREAD

**Issue by zehuanzhang** [french-paragon/BayesianMnist#1]
The ELBO consists of 3 parts: q(θ) (posterior), likelihood and p(θ) (prior).

For the likelihood, it is written in the viExperiment.py (line 90, line 103):
			loss = torch.nn.NLLLoss(reduction='mean') #negative log likelihood will be part of the ELBO
					logprob = loss(pred, labels)

For the q(θ) (posterior) and p(θ) (prior), they are in the viModel.py.
For instance, in line 159 and line 160:
		self.addLoss(lambda s : 0.5*s.getSampledWeights().pow(2).sum()/wPriorSigma**2)
		self.addLoss(lambda s : -self.out_channels/2*np.log(2*np.pi) - 0.5*s.samples['wNoiseState'].pow(2).sum() - s.lweights_sigma.sum())

Are those mean q(θ) (posterior) - p(θ) (prior)?

Looking forward to your reply!

**Comment by french-paragon** [comment]
Yes, the equation for a sample of the ELBO is given by algorithm 5 in the paper [[https://arxiv.org/pdf/2007.06823.pdf](https://arxiv.org/pdf/2007.06823.pdf)] : log(q_φ(θ)) − log(p(Dy |Dx, θ)p(θ))

Yes, the contribution of the likelyhood p(Dy |Dx, θ) is given by the loss used in viExperiment.py.

On line 159 self.addLoss(lambda s : 0.5*s.getSampledWeights().pow(2).sum()/wPriorSigma**2) add the effect of the prior p(θ), minus the log-likelyhood of the prior (up to constant) to be precise.

On line 160, the loss contribution represent the log-likelyhood of q_φ(θ) (which I prefer to call the variational posterior, instead of the posterior, as it is only an approximation of the exact posterior).

**Comment by EliseWalkerSNL** [comment]
I have some further questions on the loss function. Specifically, for the log-likelihood of $q_\phi(\theta)$, is the assumption that $q_\phi(\theta)\sim N(0,\Sigma)$, where $\Sigma$ is a trainable, diagonal matrix? If so, then line 160 in `viModel.py` should be computing evaluating the log of $N(0, \Sigma)$. If $\sigma_i$ are the diagonal entries of $\Sigma$ are there are $n$ variables, then the loss here should look like $- n/2 \log 2\pi -1/2 \sum_i \log \sigma_i  - 1/2 \sum_i x_i^2 / \sigma_i^2$. I see the first term here matching the first term in line 160 of `viModel.py`, but the second term here seems to correspond to the third term in line 160, in which case it's missing a factor of 1/2. Also, the last term here has a sum of products of the input and the $\sigma_i$, but the inputs in line 160 are not being multiplied by any $\sigma_i$. Are these typos in the loss function, or am I missing some further assumptions? Thanks!

**Comment by french-paragon** [comment]
Hello, The second term in line 160 (the last term in your equation) is the wNoiseState, this has already been normalized by the variance one line 174, in the sampleTransform function.

For the sigma, the formula for a normal distribution has a factor of 1/sqrt{2*pi*sigma^2}. You can take the sigma^2 out the square root and factor it out, leading to 1/ sqrt{2*pi} * 1/sigma. If you take the log of that you get -1/{2log{2*pi}} - log{sigma}. Here, because the variational prior is diagonal, you can separate it and you get -n/{2log{2*pi}} - Sum_i log{sigma_i}. As you see, there is no 1/2 factor.
